packages/edit-post/src/components/layout/index.js

```diff
@@ -137,16 +137,20 @@ function Layout( {
 }
 
 export default compose(
-	withSelect( ( select ) => ( {
-		mode: select( 'core/edit-post' ).getEditorMode(),
-		editorSidebarOpened: select( 'core/edit-post' ).isEditorSidebarOpened(),
-		pluginSidebarOpened: select( 'core/edit-post' ).isPluginSidebarOpened(),
-		publishSidebarOpened: select( 'core/edit-post' ).isPublishSidebarOpened(),
-		hasFixedToolbar: select( 'core/edit-post' ).isFeatureActive( 'fixedToolbar' ),
-		hasActiveMetaboxes: select( 'core/edit-post' ).hasMetaBoxes(),
-		isSaving: select( 'core/edit-post' ).isSavingMetaBoxes(),
-		isRichEditingEnabled: select( 'core/editor' ).getEditorSettings().richEditingEnabled,
-	} ) ),
+	withSelect( ( select ) => {
+		const isCodeEditingEnabled = select( 'core/editor' ).getEditorSettings().codeEditingEnabled;
+		return {
+			editorSidebarOpened: select( 'core/edit-post' ).isEditorSidebarOpened(),
+			pluginSidebarOpened: select( 'core/edit-post' ).isPluginSidebarOpened(),
+			publishSidebarOpened: select( 'core/edit-post' ).isPublishSidebarOpened(),
+			hasFixedToolbar: select( 'core/edit-post' ).isFeatureActive( 'fixedToolbar' ),
+			hasActiveMetaboxes: select( 'core/edit-post' ).hasMetaBoxes(),
+			isSaving: select( 'core/edit-post' ).isSavingMetaBoxes(),
+			isRichEditingEnabled: select( 'core/editor' ).getEditorSettings().richEditingEnabled,
+			isCodeEditingEnabled,
+			mode: isCodeEditingEnabled ? select( 'core/edit-post' ).getEditorMode() : 'visual',
```

Comment:
This doesn't seem too bad actually, but I wonder if it might be worth putting it on another PR. My reasoning being that I think the same kind of behaviour may also be wanted for `isRichEditingEnabled`.

Reply:
Yes, both should be covered as it turns out.

Reply:
so should I move this to another PR? @gziolo what do you think? I find myself not having an opinion ... :) It does introduce a new behavior and it might be good to introduce this first and then the new setting.

Reply:
Let's tackle it separately for both modes. We should also do something about the case when both are disabled 💥

Reply:
also about covering both `isCodeEditingEnabled` and `isRichEditingEnabled`, we should be aware of this behavior:

- rich editing defaults to whether the browser of the user supports Gutenberg, but it is also subject to the editor_settings WP filter
- currently the code editor is treated as a default last option, in case Gutenberg doesn't work. If we enable a behavior where we make both optional we could end up in a situation where the post opens in edit mode but no editor (visual or text) is available.

What should happen if neither visual nor text are enabled? We need to consider this situation and make a message sort of "Contact your Website Administrator" or whatever.

Reply:
Let's discuss in a separate issue. I don't have the right answer :)